src/layers/regularizers/entrywise_batch_normalization.cpp

```diff
@@ -0,0 +1,423 @@
+  El::Zero(gradient_wrt_statistics);
+  LBANN_OMP_PARALLEL_FOR
+  for (El::Int row_start = 0; row_start < local_height; row_start += bsize) {
+    const El::Int row_end = std::min(row_start + bsize, local_height);
+    const El::Int col_start = 0;
+    const El::Int col_end = local_width;
+    DataType _inv_stdev[bsize];
+    for (El::Int row = row_start; row < row_end; ++row) {
+      const auto& var = local_var(row, 0);
+      _inv_stdev[row-row_start] = 1 / std::sqrt(var + epsilon);
+    }
+    for (El::Int col = col_start; col < col_end; ++col) {
+      for (El::Int row = row_start; row < row_end; ++row) {
+        const auto& mean = local_mean(row, 0);
+        const auto& inv_stdev = _inv_stdev[row - row_start];
+        const auto& x = local_input(row, col);
+        const auto& dy = local_gradient_wrt_output(row, col);
+        auto& dmean = local_gradient_wrt_mean(row, 0);
+        auto& dvar = local_gradient_wrt_var(row, 0);
+        dmean += - dy * inv_stdev;
+        dvar += - dy * (x - mean) * inv_stdev*inv_stdev*inv_stdev / 2;
+      }
+    }
+  }
+
+  // Accumulate gradient w.r.t. statistics across processes
+  /// @todo Local statistics
+  /// @todo Arbitrary group sizes
+  comm.allreduce(gradient_wrt_statistics,
+                 gradient_wrt_statistics.RedundantComm(),
+                 El::mpi::SUM);
+
+  // Compute gradient w.r.t. input
+  //   dL/dx_i = ( dL/dy_i / sqrt(var+epsilon)
+  //             + dL/dmean / n
+  //             + dL/dvar * (x_i - mean) * 2/(n-1) )
+  const DataType inv_stats_count = DataType{1} / statistics_count;
+  const DataType inv_stats_countm1 = DataType{1} / (statistics_count - 1);
+  LBANN_OMP_PARALLEL_FOR
+  for (El::Int row_start = 0; row_start < local_height; row_start += bsize) {
+    const El::Int row_end = std::min(row_start + bsize, local_height);
+    const El::Int col_start = 0;
+    const El::Int col_end = local_width;
+    DataType _inv_stdev[bsize];
+    for (El::Int row = row_start; row < row_end; ++row) {
+      const auto& var = local_var(row, 0);
+      _inv_stdev[row-row_start] = 1 / std::sqrt(var + epsilon);
+    }
+    for (El::Int col = col_start; col < col_end; ++col) {
+      for (El::Int row = row_start; row < row_end; ++row) {
+        const auto& mean = local_mean(row, 0);
+        const auto& inv_stdev = _inv_stdev[row - row_start];
+        const auto& x = local_input(row, col);
+        const auto& dy = local_gradient_wrt_output(row, col);
+        auto& dx = local_gradient_wrt_input(row, col);
+        auto& dmean = local_gradient_wrt_mean(row, 0);
+        auto& dvar = local_gradient_wrt_var(row, 0);
+        dx = (dy * inv_stdev
+              + dmean * inv_stats_count
+              + dvar * (x - mean)) * 2 * inv_stats_countm1;
+      }
+    }
+  }
+
+}
+
+/** @brief Backprop for inference.
+ *
+ *  Assumes forward prop uses running statistics. In other words,
+ *  statistics are independent of input.
+ */
+void bp_inference_impl(DataType epsilon,
```

Comment:
"Backprop for inference" is an odd thing to say. I think this is just computing dL/dx? We should say something along those lines.

Reply:
I'd like to make a distinction between the implementation for training and for inference. Perhaps more documentation is in order.